text/0249-v2-experiments.md

```diff
@@ -0,0 +1,137 @@
+- The new @aws-cdk-previews package will take a peer dependency on @aws-cdk.
+- When a module is ready for graduation to GA, in a single commit, the code will be moved to packages/@aws-cdk and the
+  metadata tags for stability and maturity will be set to 'stable'.
+- A feature flag strategy will be devised to handle experimental APIs within stable modules, or in the core. Runtime
+  checks will make sure users who have not opted in do not call experimental APIs.
+
+# Drawbacks
+
+- When we graduate a module, users who have participated in the Developer Preview will have to update their imports to
+  make use of the stable version in the main repository.
+- For developers who are working on experimental code that has complex dependencies with the core and with other
+  modules, development will be more complicated.
+- In some cases, experimentation will not be possible. New features introduced to the main repository will have to be
+  vetted much more carefully before they are merged, since any public API released as part of v2 will have to be
+  supported for the lifetime of the major version.
+- The stable package that we vend cannot take dependencies on the experimental package.
+- API design for L2 construct libraries will have to take into account the possibility of future additions that are
+  implemented in a separate package. This might lead to awkward choices, like making functions public that otherwise
+  would have been private.
+
+# Alternatives
+
+Some alternatives to this strategy are:
+
+- Make no changes and continue to experiment in the main release.
+  - Improve documentation so that it is more obvious to users that they are importing experimental code.
+  - Make it a policy to actually include the name 'experimental' in all experimental APIs.
+- Use clever build tricks to hide experimental modules. This might be possible in Typescript, where the type files are
+  separate from implementation, but might prove difficult or impossible in other languages such as Python.
+- Call the experimental package @aws-cdk-experiments instead of @aws-cdk-previews.
+
+# Adoption Strategy
+
+The hope is that for the majority of users, upgrading from 1.x to 2.x will be as simple as changing import statements at
+the top of their source files, rebuilding, and redeploying. This upgrade should not result in any resources being
+destroyed and recreated, except in very specific cases where we have made it clear to users, such as with the EKS L2
+construct library. The same should be true for users upgrading from the preview package to the stable package as modules
+are graduated from Developer Preview to GA.
+
+# Implementation Plan
+
+This section contains examples covering common scenarios:
+
+## 1. A brand new L2, where there was only an L1 before.
+
+- Create a new package in packages/@aws-cdk-previews exactly the same we we would have for v1
```

Comment:
are we planning previews to be a single module or multiple modules? This implies that @aws-cdk-previews is an npm scope with multiple modules.

We should also make sure that we never take a dependency on experimental modules (either from the stable module or between themselves). Otherwise, we are back in peer-dependency hell again.

Reply:
A single experimental package with multiple modules.